Updated Magres parser for NMR schema

Sathya-S3 opened the pull request:
As discussed with @JFRudzinski and @EBB2675 in #7, in this PR I have completed steps 0-2 as recommended.

Quoting Joseph,
"0. migrate this repo to the FAIRmat-NFDI organization for future control and maintenance. If there is no current development at the moment, I would opt to quickly do this next week before you get started again. Please let me know if this is ok for you or if you have some potential issue.

1. remove the schema_packages folder from this repository,

2. the parser needs to change to import schema from the new package, and also reflect all new code changes to the schema package"

This PR is step 2.5 which is the formal review process to incorporate changes into the central repository main branch.

**Highlights of this PR:**
1. Project dependencies updated in pyproject.toml -> import new NMR schema v0.1.0
2. workflow.py moved from old schema_packages folder within src to parsers folder to preserve existing CASTEP workflow and continue development on including more DFT code workflows after this PR.
3. schema_packages folder and contents removed.
4. Parser updates:
    - NMR quantities imported from [nomad-schema-plugin-nmr](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr). Data quantity classes added and decomposed contributions for EFG (local and non-local) removed as per the new NMR schema. Indirect spin-spin coupling contributions alogn with the total contribution introduced.
    - Model System and Atoms State class references changed to align with nomad-simulations data model updates (v0.4.0). atom_states changed to particle_states; positions (from [atoms][/atoms] block) moved from model_system.cell to model_system.
    - regex and logic for parsing kpoint_mp_offset info from [calculation][/calculation] block changed to allow for spaces between entries and potentially trailing spaces, for increased success of parsing.
    - Lookup tables introduced for particle states and particle state pairs using indices from magres files for accurate pairing of data from magres files to custom named labels in the database. Helper functions introduced to successfully complete the above operation with code repeatability.
    - Transposing of tensors removed, as it will result in different derived quantities for some classes when normalizing using schema definitions.
    - Quantity units made consistent with new NMR schema.

Sathya-S3 committed:
pyproject edits

Sathya-S3 committed:
Remove schema files from package

Sathya-S3 committed:
moved workflow.py to parser folder

Sathya-S3 committed:
imports change

Sathya-S3 committed:
'atom' block regex update

Sathya-S3 committed:
class name updates, additions

Sathya-S3 committed:
parse positions, particle_states correctly

Sathya-S3 committed:
added new helper functions

Sathya-S3 committed:
mag shielding parsing updates

Sathya-S3 committed:
electric field gradient parsing updates

Sathya-S3 committed:
indirect spin couplings parsing addition, updates

Sathya-S3 committed:
mag sus parsing function updates

Sathya-S3 committed:
parse outputs function updates

Sathya-S3 committed:
parse function updates

Sathya-S3 committed:
added global variables for particle states lookup

Sathya-S3 committed:
ruff quote style change

Sathya-S3 committed:
schema package version change

Sathya-S3 committed:
kpoint parsing changes

Sathya-S3 committed:
mag sus scale property removed

Sathya-S3 committed:
test parser and data change

Sathya-S3 committed:
corrected index creation

Sathya-S3 committed:
corrected index and chem symbols in tests

Sathya-S3 committed:
added custom labels to test magres file

Sathya-S3 committed:
ruff linting fixes

Sathya-S3 committed:
reverted moved imports

Sathya-S3 committed:
removed debug messages

Sathya-S3 committed:
removed transposing

Sathya-S3 committed:
removed transpose from tests

Sathya-S3 commented:
@EBB2675 Wanted to ask a couple of questions about how the NMR quantities are expressed on the GUI. 

1. I tested with my nomad-distro-dev and found that when extracted, the quantity units with meters are converted to Angstrom, J converted to eV, and similar. Is there a reason why this happens on the GUI? If this avoidable so that the NMR quantities are expressed just using the units defined in the parser + schema?
2. Is there a reason why the tensors when displayed on the GUI are transposed by default? There were some adjustments in the older parser version, where for the purpose of displaying tensors correctly, they were transposed. But now in the NMR schema, we are using the tensor values to also further calculate a few derived quantities which may change in value if transposed especially when it's not a symmetric tensor. So, I have taken the transposes out of the code now.

Many thanks for your time and advice.

Sathya-S3 commented:
@jkshenton FYI

EBB2675 commented:
hi @Sathya-S3, just wanted to quickly give you a heads-up that i am looking into it and will return to you tomorrow

EBB2675 commented:
Hi again @Sathya-S3 

i have processed the `ethanol_nmr.magres` file in the tests folder.

1) The units can be switched in the GUI via the UNITS dropdown on the top right. For instance, one can select Dimension as Energy, and change the default unit (please see the attached screenshot)

<img width="1362" height="943" alt="Screenshot from 2025-07-16 17-07-04" src="https://github.com/user-attachments/assets/cbb1e416-4a26-481a-8f51-d640d29cd9a8" />

EDIT: See below for the tensor problem

Sathya-S3 commented:
@EBB2675 Thank you for the clarifications. 

1. For the units, I can confirm that when I choose SI units, the units are displayed correctly as we would expect. Screenshot below:

<img width="1607" height="840" alt="nomad_units_screenshot" src="https://github.com/user-attachments/assets/e4484058-690a-4c72-ba72-4992b0a20680" />

2. @jkshenton Please read Esma's comment above. As you work a lot more with CASTEP outputs, could you please confirm if Fortran tensor outputs are in column-major order, and therefore, in need of transposing?

Sathya-S3 commented:
@EBB2675 Following discussion with @jkshenton, we have started a conversation with one of the lead CASTEP developers in CCP-NC to clarify on how CASTEP writes the tensors out to the magres files (row-major or column-major).

But, I think my question about transposing it for NOMAD GUI needs to be treated independent of the row-major situation, as my query pertains to what the GUI does to the tensor only. I'll try to explain in detail below:

1. **What the magres specification says (vs) magres parser code:**
- The [magres specification document](https://www.ccpnc.ac.uk/docs/magres/magres-format.pdf) (see page 5 - section 4.5.3 under ms - Magnetic Shielding), which specifies how the flattened tensor indices in magres relate to the original 3x3 NMR tensor. Screenshot below as well:
<img width="587" height="167" alt="image" src="https://github.com/user-attachments/assets/c679ed1d-9005-4463-86d3-4d0b526a67fa" />

- To extract this tensor correctly, a simply numpy reshape should work, which is what we do in the magres code - this should make the tensor a row-major order as you mentioned how the table widget expects. Please see [parse_magnetic_shieldings](https://github.com/FAIRmat-NFDI/nomad-parser-magres/blob/e7cf359c8b84090e375c76789e5d9be632dbbd9e/src/nomad_parser_magres/parsers/parser.py#L668) and [extract_particle_state_and_tensor](https://github.com/FAIRmat-NFDI/nomad-parser-magres/blob/e7cf359c8b84090e375c76789e5d9be632dbbd9e/src/nomad_parser_magres/parsers/parser.py#L634) functions.

2. **Magres parser (vs) GUI:**
- We created a dummy basic magres file for a 2-atom system named test_transpose2.magres (please see attached file -  inside the zip) 
[test_transpose2.zip](https://github.com/user-attachments/files/21485487/test_transpose2.zip). The first row of magres magnetic shieldign for atom 1 has similar numbers to the magres specification example, just scaled by 1E+6 because it will be scaled down by 1E-6 in the code anyway. I first used the NOMAD cli to run the parser on this file and extract the outputs to a json file (see output_test_transpose.json attached)
[output_test_transpose.json](https://github.com/user-attachments/files/21485499/output_test_transpose.json). The parser code does not transpose 'values' field which contains the reshaped tensor. In the outputs section of the json file, for each atom, you can see that the relationship between the reshaped magnetic shielding tensor and the actual flat tensor in the attached magres file are the same as in the magres specification section quoted in point 1.
- When I boot up the GUI for my localhost copy that reads from the same parser code, however, I see the tensor displayed as a transposed version. Please see the screenshot below showing the tensor for atom 1, appearing transposed in the GUI:
<img width="1460" height="932" alt="test_transpose2_no_transpose_in_parser" src="https://github.com/user-attachments/assets/948e762f-d4d0-447f-99b5-12cce11dec1a" />

3. We already brought this issue up when the magres parser was initially added as part of the electronicparsers. For your reference, see [issue #171](https://github.com/nomad-coe/electronic-parsers/issues/171#issuecomment-2022918802) - especially my comment posted on Feb 27, 2024 and Jose's reply on dates Feb 27, 2024 and Mar 27, 2024 (when he fixed it). The fix is in commit id [8220643](https://github.com/nomad-coe/electronic-parsers/commit/82206430661c84bcfd9ab4a836d3458310c31dbc) where np.transpose was added to the tensors when parsing. While this was a fix to display the tensor correctly on the GUI, we don't think it is the right approach to transpose the tensor and does not directly address why the table widget transposes the tensor in the first place.

Would this information be sufficient to address why the table widget transposes the tensor when it needn't do it?

Sathya-S3 commented:
**Additional context:**
When I uploaded the test_transpose2.magres file to nomad-lab website which still uses the electronicparsers package for the magres parser entrypoint, you can see that the tensor appears the same in the GUI as it does in the json file I shared in the previous message. This is expected, following Jose's addition of transposing mentioned in my previous comment. Screenshot below for reference:
<img width="1913" height="991" alt="test_transpose2_transpose_in_electronicparsers" src="https://github.com/user-attachments/assets/2fffcfc0-8a7d-4ae9-b508-0f40435cd6e7" />

EBB2675 commented:
thank you very much @Sathya-S3, now I have the necessary context :thumbsup:

I am checking the source code and will return to you shortly

EBB2675 commented:
Hi again @Sathya-S3 

i have found the relevant piece of source code, and i have spoken with the development team. They agreed on fixing the transpose in the GUI issue asap, probably will make it to the next release, FYI

Sathya-S3 commented:
> Hi again @Sathya-S3
> 
> i have found the relevant piece of source code, and i have spoken with the development team. They agreed on fixing the transpose in the GUI issue asap, probably will make it to the next release, FYI

@EBB2675 That's fantastic! Thank you very much for addressing the issue so quickly.

Both my questions have now been answered, thanks again. So, as long as you and any other reviewers that you may choose to invite to review this PR are happy with the latest additions to the magres parser, I don't have anything else to add before merging to the main branch.

Sathya-S3 committed:
Remove stale submodule reference for nomad-nmr-schema

Sathya-S3 committed:
nmr schema dependency edit

Sathya-S3 committed:
Add Jacob's ladder classification

Sathya-S3 committed:
removed repeated code line

Sathya-S3 committed:
schema dependency version change

Sathya-S3 committed:
Updated regex - debug failed parsing

Sathya-S3 committed:
Updated python version in pyproject

Sathya-S3 committed:
nmr schema commit id updated

Sathya-S3 committed:
added support for QE magres parsing

Sathya-S3 committed:
improved qe version filtering

Sathya-S3 committed:
update parser to pass indices for site name resolution

Sathya-S3 committed:
linting fix

Sathya-S3 committed:
nmr schema commit id update

Sathya-S3 committed:
updates to site label parsing + tests

Sathya-S3 committed:
MS, EFG pint units update, remove scaling

Sathya-S3 committed:
parser test update with units change and removed scaling

Sathya-S3 committed:
nmr schema commit id update

Sathya-S3 commented:
@JFRudzinski @ladinesa Please note, there have been changes to this branch since the last PR summary/description which was a while ago. For your reference, when reviewing.

### **UPDATES (post-July 16 2025 / after commit [e7cf359](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/e7cf359c8b84090e375c76789e5d9be632dbbd9e))**

**Schema dependency maintenance:**
- Multiple updates to the pinned NMR schema commit/version to track upstream changes and ensure compatibility with evolving NOMAD schema releases ([b61f42d](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/b61f42d4f0c42ab35ed4367214e328bd41c19cc0), [214918e](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/214918e1a4a22d282b1b717f68770aeeaca9336b), [e33c342](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/e33c342be8b2b530474fcbc05900a5932607e95d), [217c2e0](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/217c2e0fd5cd9eeedfc038bd02552c716b008d41), [eb28d64](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/eb28d641df021bd907d1c0b2d73486a30fd624b7)).
- Removed a stale submodule reference to nomad-nmr-schema, ensuring the parser depends solely on the maintained schema plugin ([9570e67](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/9570e675d47b25f78edd85e10829e29b0b933719)).

**Unit handling and tensor normalization:**
- Updated magnetic shielding and EFG quantities to use Pint units directly, removing internal scaling factors during parsing ([6fafae8](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/6fafae82096a882c753d6bf84f583d5b607b7670)).
- Parser tests updated accordingly to reflect the new unit handling and removal of implicit 1e-6 scaling ([e1daed3](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/e1daed3d525dfafd866a0e66e2a411f7feaf8284)).

**Quantum ESPRESSO MAGRES support:**
- Added full parsing support for QE-generated MAGRES files ([1c47825](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/1c47825c309c6caf5aac8d38d0a346b3f97350b8)).
- Improved QE version filtering logic for more robust workflow detection ([0c5cf73](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/0c5cf73118b91d2c4bf0e6d699a8b9c9ba8e055e)).
- Regex improvements to fix edge cases where parsing previously failed ([1c52a9d](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/1c52a9d5b042e43dfb2500a6a7b889b9aec290c4)).

**Site label and atom-state mapping improvements:**
- Updated parser logic to pass explicit particle/site indices for reliable site-name resolution ([99f88bf](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/99f88bfe97701a1a752b279b3db2e0ad629016b5)).
- Improvements to site-label parsing and corresponding tests ([9caf01b](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/9caf01b63e98a666174a0b422ffb386ddd2e8713)).

**Metadata enrichment:**
- Added Jacob’s-ladder functional classification to parsed metadata for improved method categorisation in NOMAD ([dbe52d8](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/dbe52d88acae564410dc12e2c59cb99fe1d04443)).

**General maintenance and hygiene:**
- Updated the required Python version in pyproject.toml ([3ae9bf5](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/3ae9bf5004701f4459bd5c5496de077901cfbbfb)).
- Minor linting and cleanup fixes ([a0114fb](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/a0114fb1b7c014e92de5a47795ef730f9bbb0f28), [36f6519](https://github.com/FAIRmat-NFDI/nomad-parser-magres/pull/10/commits/36f6519446178050ed9669db73de9c1d7cccfbb2)).

@jkshenton FYI

Sathya-S3 committed:
EFG Hartree atomic unit

JFRudzinski commented:
Hi @Sathya-S3 ! My apologies for the long delay.

I have discussed with @ladinesa and for consistent maintainability of this plugin within the public deployment, we would like to integrate it into our [simulation parsers package](https://github.com/FAIRmat-NFDI/nomad-parser-plugins-simulation). 

For future development from your side, you will be able to simply fork that repo, develop, put in PR with upstream repo. This way you don't have to worry about also maintaining compatibility in terms of nomad-lab or other dependencies changing, we will take care of that.

Let me know if you have any further questions about this process, or if you think the integration will be somehow problematic for you -- I imagine there may be some overhead for you if you want to disable many of the parser entry points from our repo when you install in your Oasis ... we can discuss this.

In case we agree to this path: To move forward then, we simply need to bring this repo into a stable state, including completing and merging (or closing) any open PRs. I think both open issues are not relevant anymore then. 

In terms of the current code changes, you can simply fix the tests and then merge the PR. We will need to manually transfer the code to our repo, at which time we will do a technical code review and refinement.

Sathya-S3 commented:
> Hi @Sathya-S3 ! My apologies for the long delay.
> 
> I have discussed with @ladinesa and for consistent maintainability of this plugin within the public deployment, we would like to integrate it into our [simulation parsers package](https://github.com/FAIRmat-NFDI/nomad-parser-plugins-simulation).
> 
> For future development from your side, you will be able to simply fork that repo, develop, put in PR with upstream repo. This way you don't have to worry about also maintaining compatibility in terms of nomad-lab or other dependencies changing, we will take care of that.
> 
> Let me know if you have any further questions about this process, or if you think the integration will be somehow problematic for you -- I imagine there may be some overhead for you if you want to disable many of the parser entry points from our repo when you install in your Oasis ... we can discuss this.
> 
> In case we agree to this path: To move forward then, we simply need to bring this repo into a stable state, including completing and merging (or closing) any open PRs. I think both open issues are not relevant anymore then.
> 
> In terms of the current code changes, you can simply fix the tests and then merge the PR. We will need to manually transfer the code to our repo, at which time we will do a technical code review and refinement.

Thank you @JFRudzinski and @ladinesa. Happy to see NOMAD developments moving along and to see our contributions to magres parser being ready to be integrated to the official simulation parsers.

We're okay with the proposed approach of contributing in the future. I don't mind the mild overhead to selectively disable some entry points.

Regarding the stability of the repo's current state, I think some help from Nomad side would be great! We have made a unit change to the quantity called electric field gradient - instead of expressing it in SI units V/m^2, we have changed it to be expressed as a single atomic unit. This change ahs been accepted into pint's master branch as well: https://github.com/hgrecco/pint/commit/40936b9441156a8e0260034126a93b553882077b

The test that is currently failing on our schema and parser repos is because Nomad uses it's local version of pint's unit registry and the unit change is not reflected in the official nomad-lab dependency release yet. I have made @lauri-codes aware of this and he has made a point to included these changes in the future nomad-lab release v1.4.2. But, since this is not due for a month at least and we are pushing for our Oasis release this month, Lauri and I agreed that as an interim solution, I create a forked copy of nomad-lab with these custom changes so that our tests pass and our Oasis is stable. I'll merge the PR once these fixes are in place. Apologies for the slight delay and thank you for your patience.

JFRudzinski commented:
@Sathya-S3 sounds good, thanks for the update. I am glad we are approaching a good longer-term collaborative solution! 

I will keep a look out when the fix is in, but feel free to remind me when action is needed

Sathya-S3 committed:
nomad-lab dependency patch

Sathya-S3 committed:
nmr schema commit id update

Sathya-S3 committed:
dependencies change

Sathya-S3 committed:
temporarily bypass test to prevent 'a_u_efg' missing error

Sathya-S3 committed:
skipped test until 'a_u_efg' added to nomad-lab units

Sathya-S3 committed:
nmr schema ms units update

Sathya-S3 committed:
dependency updates, reenabled pytest
